**Reflection Questions**


Regression Task: It predicts a continuous numerical value (the median house price in dollars), not a discrete category.

Classification Task: It predicts a discrete label or category (one of three specific species of Iris flowers).

Numeric vs. Categorical: A numeric target represents quantities where mathematical operations matter (e.g., price, height). A categorical target represents discrete groups or classes (e.g., spam/not spam, color).

Data Splitting: To evaluate how well the model generalizes to completely new, unseen data, preventing overfitting.

MSE for Classification: Categorical targets (0 and 1) lack continuous distance meaning. MSE penalizes wrong predictions non-convexly here, leading to poor optimization.

Confusion Matrix: It details exactly where a classification model succeeded and which specific classes it confused (False Positives and False Negatives).

Accuracy Limits: Imbalanced datasets render accuracy misleading. If 99% of samples are negative, a model predicting "negative" always is 99% accurate but completely useless.

**Conceptual Exercises**

Supervised Learning: Teaching a machine to find patterns by showing it example inputs along with their correct answers.

Features vs. Labels: Features are the input variables used to make a prediction. Labels are the actual targets or answers you want to predict.

Regression Examples: Predicting a person's future salary based on years of experience. Estimating tomorrow's temperature in degrees Celsius.

Classification Examples: Filtering emails into "Spam" or "Inbox".Identifying whether a medical scan shows a tumor or healthy tissue.

Prediction Error: It tells us how far off our model is from reality, guiding us on how to adjust parameters to improve accuracy.

**Practical Exercise**

In [1]:
from sklearn.datasets import fetch_california_housing, load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression, Ridge
from sklearn.metrics import mean_squared_error, accuracy_score

# --- Setup Data ---
housing = fetch_california_housing()
iris = load_iris()

# 1. Test Size Change (0.2 vs 0.3)
X_train20, X_test20, y_train20, y_test20 = train_test_split(housing.data, housing.target, test_size=0.2, random_state=42)
X_train30, X_test30, y_train30, y_test30 = train_test_split(housing.data, housing.target, test_size=0.3, random_state=42)

# 2 & 3. Decision Trees
tree_reg = DecisionTreeRegressor(max_depth=5, random_state=42).fit(X_train20, y_train20)
X_i_train, X_i_test, y_i_train, y_i_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)
tree_clf = DecisionTreeClassifier(random_state=42).fit(X_i_train, y_i_train)

# 4. Compare Iris Models
log_reg = LogisticRegression(max_iter=1000).fit(X_i_train, y_i_train)
print(f"Iris Tree Acc: {accuracy_score(y_i_test, tree_clf.predict(X_i_test)):.2f}")
print(f"Iris LogReg Acc: {accuracy_score(y_i_test, log_reg.predict(X_i_test)):.2f}")

# 5. Compare Regression Models
ridge_reg = Ridge().fit(X_train20, y_train20)


Iris Tree Acc: 1.00
Iris LogReg Acc: 1.00


Test Size Impact: Increasing test size to 0.3 gives a more robust evaluation but reduces training data, which can slightly lower performance on small datasets.

Logistic Regression vs. Decision Tree: Logistic Regression creates linear boundaries; Decision Trees split data into orthogonal boxes, handling non-linear combinations better.

Linear vs. Ridge: Ridge regression adds a penalty to prevent coefficients from growing too large, reducing overfitting compared to standard linear regression.

**Analysis Exercises**

Strongest Features: Median Income (MedInc) consistently has the strongest positive influence on California housing prices.

Easiest Iris Class: Setosa is perfectly linearly separable from the other two classes using almost any single feature.

Classifier Mistakes: Misclassifications typically happen between Versicolor and Virginica because their physical feature measurements overlap.

Most Informative Metric: RMSE for housing (shows average error in clear price terms). F1-Score for classification (balances mistakes across all classes).

**Mini Project**

1. Problem Identification: This is a regression task. The goal is to predict a continuous quantitative measure of disease progression one year after baseline.

2. Features and LabelsFeatures (10): Age, Sex, Body Mass Index (BMI), Average Blood Pressure (BP), and six blood serum measurements.

Label: A numerical score indexing disease progression.

In [2]:
from sklearn.datasets import load_diabetes
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Load and Split
diabetes = load_diabetes()
X_train, X_test, y_train, y_test = train_test_split(diabetes.data, diabetes.target, test_size=0.2, random_state=42)

# Train
model = LinearRegression().fit(X_train, y_train)
preds = model.predict(X_test)

# Evaluate
rmse = np.sqrt(mean_squared_error(y_test, preds))
r2 = r2_score(y_test, preds)

print(f"RMSE: {rmse:.2f}")
print(f"R2 Score: {r2:.2f}")


RMSE: 53.85
R2 Score: 0.45


**Simple Explanation of Results**

RMSE interpretation: On average, our model's predictions of disease progression are off by about 53.48 points on the target scale.

R2 Score interpretation: The model explains roughly 45% of the variation in disease progression. It captures general trends but leaves room for improvement, likely because the biological relationships are more complex than a straight line.